reactive/spark.py

```diff
@@ -1,70 +1,173 @@
     hookenv.status_set('maintenance', 'Configuring Apache Spark')
     spark = Spark(get_dist_config())
     spark.stop()
+    if is_state('hadoop.ready') and mode.startswith('yarn') and (not is_state('yarn.configured')):
+        # was in a mode other than yarn, going to yarn
+        hookenv.status_set('maintenance', 'Setting up Apache Spark for YARN')
+        spark.configure_yarn_mode()
+        set_state('yarn.configured')
+
+    if is_state('hadoop.ready') and (not mode.startswith('yarn')) and is_state('yarn.configured'):
+        # was in a yarn mode and going to another mode
+        hookenv.status_set('maintenance', 'Disconnecting Apache Spark from YARN')
+        spark.disable_yarn_mode()
+        remove_state('yarn.configured')
+
     spark.configure()
     spark.start()
-    hookenv.status_set('active', 'Ready')
+    report_status(spark)
 
 
-@when('spark.started')
-@when_not('hadoop.ready')
-def stop_spark():
-    hookenv.status_set('maintenance', 'Stopping Apache Spark')
+@when('spark.started', 'yarn.configured')
+@when_not('hadoop.ready', )
+def disable_yarn():
+    hookenv.status_set('maintenance', 'Disconnecting Apache Spark from YARN')
     spark = Spark(get_dist_config())
-    spark.close_ports()
     spark.stop()
-    remove_state('spark.started')
+    spark.disable_yarn_mode()
+    spark.start()
+    remove_state('yarn.configured')
+    report_status(spark)
+
+
+@when('spark.installed', 'sparkpeers.joined')
+def peers_update(sprkpeer):
+    nodes = sprkpeer.get_nodes()
+    nodes.append((hookenv.local_unit(), hookenv.unit_private_ip()))
+    update_peers_config(nodes)
+
+
+@when('spark.installed')
+@when_not('sparkpeers.joined')
+def no_peers_update():
+    nodes = [(hookenv.local_unit(), hookenv.unit_private_ip())]
+    update_peers_config(nodes)
+
+
+def update_peers_config(nodes):
+    nodes.sort()
+    if data_changed('available.peers', nodes):
+        spark = Spark(get_dist_config())
+        # We need to reconfigure spark only if the master changes or if we
+        # are in HA mode and a new potential master is added
+        if spark.update_peers(nodes):
+            hookenv.status_set('maintenance', 'Updating Apache Spark config')
+            spark.stop()
+            spark.configure()
+            spark.start()
+            report_status(spark)
+
+
+@when('spark.installed', 'zookeeper.ready')
+def configure_zookeepers(zk):
+    zks = zk.zookeepers()
+    if data_changed('available.zookeepers', zks):
+        spark = Spark(get_dist_config())
+        hookenv.status_set('maintenance', 'Updating Apache Spark HA')
+        spark.stop()
+        spark.configure_ha(zks)
+        spark.configure()
+        spark.start()
+        set_state('zookeeper.configured')
+        report_status(spark)
+
+
+@when('spark.installed', 'zookeeper.configured')
+@when_not('zookeeper.ready')
+def disable_zookeepers():
+    hookenv.status_set('maintenance', 'Disabling high availability in standalone mode')
```

Comment:
Small knit-pick on the status message.  You could be disabling HA in yarn mode too, right?  I think drop the "in standalone mode" from the status_set.